### PR TITLE
Fix possibly unaligned writes in MutableBuffer

### LINKED ISSUE #1410: Possible unaligned write with MutableBuffer::push
The following test fails when run under miri:

```
    #[test]
    fn mutable_push_unaligned_u64() {
        let mut buf = MutableBuffer::new(16);
        buf.push(1_u8);
        buf.push(1_u64);
        assert_eq!(9, buf.len());
        assert_eq!(&[1u8, 1u8, 0, 0, 0, 0, 0, 0, 0], buf.as_slice());
    }
```

The code uses `std::ptr::write` to append the u64, regardless of the current alignment. Replacing that with `write_unaligned` would probably work. The type parameter is constrained by the `ToByteSlice` trait, but it seems the trait is not actually used. Not sure whether this is expected. It seems `ToByteSlice` is more of a marker trait here that indicates that the contents can be transmuted to a slice of bytes.

### PR DESCRIPTION
**Pull request by jhorstmann**
# Which issue does this PR close?

Closes #1410.

# Rationale for this change

Changes the `push`, `push_unchecked` and `extend_from_iter` functions so that data is transfered using `copy_nonoverlapping` which works for unaligned destinations. LLVM will most likely transform this again into a single unaligned write. This also makes these methods actually use the `ToByteSlice` trait. In `extend_from_slice`, which this PR did not change, the trait is more used like a marker trait because the whole slice is copied without going through the trait methods. This is probably fine since it's supposed to be an internal trait that is only implemented for `ArrowNativeType`.

The `from_trusted_len_iter` did not have a problem with alignment because the buffer is created inside the method, but for consistency I also changed them to use `to_byte_slice` and `copy_nonoverlapping`.

All newly added tests were previously failing when running under miri.

### CODE REVIEW

**Review by sunchao — COMMENTED** [review]
Since we have this for `extend_from_iter`, do we need the fix for `from_trusted_len_iter` as well?

**Comment by jhorstmann** [comment]
In `from_trusted_len_iter` the `MutableBuffer` is created inside the method, so there should be no way for the pointer to become unaligned. Probably still a good idea to adjust it for consistency.

**Review by sunchao — APPROVED** [review]
Ah you are right. LGTM

File: arrow/src/buffer/mutable.rs [hunk 1]
```diff
@@ -710,6 +713,46 @@ mod tests {
...
+        assert_eq!(9, buf.len());
+        assert_eq!(&[1u8, 1u8, 0, 0, 0, 0, 0, 0, 0], buf.as_slice());
+    }
+
+    #[test]
+    fn mutable_extend_from_slice_unaligned_u64() {
+        let mut buf = MutableBuffer::new(16);
+        buf.extend_from_slice(&[1_u8]);
+        buf.extend_from_slice(&[1_u64]);
+        dbg!(&buf.as_slice());
```
Comment (sunchao) [inline arrow/src/buffer/mutable.rs:731]: nit: is this needed?

Comment (jhorstmann) [reply arrow/src/buffer/mutable.rs]: Thanks for the review! The dbg is now removed.

**Comment by sunchao** [comment]
Merged, thanks!

**Pull request merged**